File: yujung/프로그래머스/더맵게.cpp
#include <string>
#include <vector>
#include<queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    priority_queue<int,vector<int>, greater<int>> pq;
    int answer = 0;
    for(auto a: scoville){
        pq.push(a);
    }
    int sum=0;
    while(pq.size()>=2&&pq.top()<K){
        sum=0;
        int first=pq.top();
        pq.pop();
        int second=pq.top();
        pq.pop();
        sum=first+second*2;
        pq.push(sum);
        answer++;
    }
   if(!pq.empty()&&pq.top()<K) return -1;
    return answer;
}

File: seoyeon/프로그래머스/알고리즘 고득점 Kit/힙(Heap)/더맵게.py
# 1.내 코드 
# 정확성: 71.0 (테케 1,3,8,14 실패)
# 효율성: 16.1
# 합계: 87.1 / 100.0
import heapq

def solution(scoville, K):
    answer = 0
    n = len(scoville)
    
    heapq.heapify(scoville) 
    while True:
        
        if scoville[0]>=K:
            break
        if answer>n:
            answer = -1
            break
        answer += 1
                
        x=heapq.heappop(scoville)
        y=heapq.heappop(scoville)
        
        heapq.heappush(scoville,x+2*y)
        
    return answer

#2. 정답
import heapq

def solution(scoville, K):
    answer = 0
    n = len(scoville)
    
    heapq.heapify(scoville) 
    while True:
        
        if scoville[0]>=K:
            break
        if len(scoville)<2: #내 코드와 이 부분 다름
            answer = -1
            break
        answer += 1
                
        x=heapq.heappop(scoville)
        y=heapq.heappop(scoville)
        
        heapq.heappush(scoville,x+2*y)
        
    return answer